File: frontend (React)/src/services/BookService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/books';

class BookService {
  getAllBooks() {
    return axios.get(API_URL);
  }

  getBookById(id) {
    return axios.get(`${API_URL}/${id}`);
  }

  createBook(book) {
    return axios.post(API_URL, book);
  }

  updateBook(id, book) {
    return axios.put(`${API_URL}/${id}`, book);
  }

  deleteBook(id) {
    return axios.delete(`${API_URL}/${id}`);
  }

  searchBooks(query) {
    return axios.get(`${API_URL}/search?query=${query}`);
  }
  
  searchByTitle(title) {
    return axios.get(`${API_URL}/search/title?title=${title}`);
  }

  searchByAuthor(author) {
    return axios.get(`${API_URL}/search/author?author=${author}`);
  }

  searchByIsbn(isbn) {
    return axios.get(`${API_URL}/search/isbn?isbn=${isbn}`);
  }
}

export default new BookService();

File: backend (SpringBoot)/src/main/java/com/induwara/librarymanagement/service/BookService.java
package com.induwara.librarymanagement.service;

import com.induwara.librarymanagement.dto.BookSearchDTO;
import com.induwara.librarymanagement.model.Book;

import java.util.List;

public interface BookService {

    List<Book> getAllBooks();

    Book getBookById(Long id);

    Book createBook(Book book);

    Book updateBook(Long id, Book book);

    void deleteBook(Long id);

    List<Book> searchBooks(String searchTerm);

    List<Book> findByTitle(String title);

    List<Book> findByAuthor(String author);

    Book findByIsbn(String isbn);
    
    // New method for advanced search using DTO
    List<Book> advancedSearch(BookSearchDTO searchDTO);
}